src/status_list/mod.rs

```diff
@@ -0,0 +1,504 @@
+        let status_list = encoded_list.decode_decompress()?;
+        Ok(StatusList {
+            status_size: Bits::try_from(encoded_list.status_size)?,
+            status_list,
+            aggregation_uri: encoded_list.aggregation_uri,
+        })
+    }
+}
+
+impl StatusList {
+    /// Takes a Vec of Statuses, each represented by a u8 and packs them into a byte according to the status size.
+    /// A status size of 8 bits would therefore need no packing.
+    /// A status size of 1 bit would pack 8 statuses into a single byte.
+    pub fn pack_statuses_into_bytes(
+        &mut self,
+        statuses: Vec<StatusType>,
+    ) -> Result<Vec<u8>, OAuthTSLError> {
+        let status_size_usize = self.status_size as usize;
+        let statuses_per_byte = 8 / status_size_usize;
+        let mut packed: Vec<u8> = Vec::with_capacity(statuses.len().div_ceil(statuses_per_byte));
+        let mut byte: u8 = 0;
+        let mut bit_pos = 0;
+
+        for status in statuses.into_iter() {
+            if status as usize >= (1 << status_size_usize) {
+                return Err(OAuthTSLError::InvalidStatusType(status as u8));
+            }
+
+            byte |= (status as u8) << bit_pos;
+            bit_pos += status_size_usize;
+
+            if bit_pos >= 8 {
+                packed.push(byte);
+                byte = 0;
+                bit_pos = 0;
+            }
+        }
+
+        // Flush the last byte
+        if bit_pos > 0 {
+            packed.push(byte);
+        }
+
+        self.status_list = packed.clone();
+
+        Ok(packed)
+    }
+
+    pub fn get_index(&self, index: usize) -> Result<u8, OAuthTSLError> {
+        let status_list_len = self.status_list.len() * (8 / self.status_size as usize);
+        if index >= status_list_len {
+            return Err(OAuthTSLError::IndexNotFound(index));
+        }
+
+        let byte = self.status_list[index * self.status_size as usize / 8];
+        let bit_index = (index * self.status_size as usize) % 8;
+        match self.status_size {
+            Bits::One => {
+                let mask = 0b1 << bit_index;
+                Ok((byte & mask) >> bit_index)
+            }
+            Bits::Two => {
+                let mask = 0b11 << bit_index;
+                Ok((byte & mask) >> bit_index)
+            }
+            Bits::Four => {
+                let mask = 0b1111 << bit_index;
+                Ok((byte & mask) >> bit_index)
+            }
+            Bits::Eight => {
+                let mask = 0b11111111 << bit_index;
+                Ok((byte & mask) >> bit_index)
+            }
+        }
+    }
+
+    /// Sets the status at the specified index to the given value.
+    /// Always enlarges the status list to the required size if it is not already large enough.
+    /// Therefore an index can never be out of bounds.
+    pub fn set_index(&mut self, index: usize, value: u8) -> Result<(), OAuthTSLError> {
```

Comment:
Would this perhaps make sense instead?: `fn set_status(&mut self, index: usize, status: StatusType)`